src/v/kafka/server/handlers/alter_partition_reassignments.cc

```diff
@@ -0,0 +1,259 @@
+    topic_response.partitions = std::move(invalid_partitions);
+    resp_it = std::move(topic_response);
+
+    return valid_partitions_end;
+}
+
+template<>
+ss::future<response_ptr> alter_partition_reassignments_handler::handle(
+  request_context ctx, [[maybe_unused]] ss::smp_service_group ssg) {
+    alter_partition_reassignments_request request;
+    request.decode(ctx.reader(), ctx.header().version);
+    log_request(ctx.header(), request);
+    alter_partition_reassignments_response resp;
+
+    if (!ctx.authorized(
+          security::acl_operation::alter, security::default_cluster_name)) {
+        resp.data.error_code = error_code::cluster_authorization_failed;
+        resp.data.error_message = ss::sstring{
+          error_code_to_str(error_code::cluster_authorization_failed)};
+        co_return co_await ctx.respond(std::move(resp));
+    }
+
+    // TODO(@NyaliaLui) - Check if we are the controller
+
+    resp.data.responses.reserve(request.data.topics.size());
+    auto all_node_ids = ctx.metadata_cache().node_ids();
+
+    for (reassignable_topic& topic : request.data.topics) {
+        reassignable_topic_response topic_response{.name = topic.name};
+
+        auto valid_partitions_end = validate_partitions(
+          topic.partitions.begin(),
+          topic.partitions.end(),
+          std::back_inserter(resp.data.responses),
+          topic_response,
+          all_node_ids);
+
+        for (auto it = topic.partitions.begin(); it != valid_partitions_end;
+             ++it) {
+            const reassignable_partition& partition{*it};
+            model::ntp ntp{
+              model::kafka_namespace, topic.name, partition.partition_index};
+
+            if (partition.replicas.has_value()) {
+                auto assignments
+                  = co_await ctx.topics_frontend().generate_reassignments(
+                    partition.partition_index, topic.name, *partition.replicas);
+                if (
+                  assignments.has_value()
+                  && assignments.value().front().id
+                       == partition.partition_index) {
+                    std::vector<model::broker_shard> new_replicas
+                      = assignments.value().front().replicas;
+                    vlog(
+                      klog.debug,
+                      "Request to reassign partitions: ntp {}, replicas {}",
+                      ntp,
+                      new_replicas);
+
+                    auto errc
+                      = co_await ctx.topics_frontend().move_partition_replicas(
+                        ntp,
+                        new_replicas,
+                        model::timeout_clock::now() + request.data.timeout_ms);
+
+                    if (!errc) {
+                        topic_response.partitions.push_back(
+                          reassignable_partition_response{
+                            .partition_index = partition.partition_index});
+                    } else {
```

Comment:
there are generally a lot of conditions in this handler that result in information being returned to the client. that's nice to have for the clients, but some cilents may toss away that information and not present it. therefore it's really quite useful for us to also log information about these conditions to make debugging easier. they can be logged at debug (even the errors that are returned to the users).